config/tls_test.go

```diff
@@ -0,0 +1,179 @@
+package config
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"math/big"
+	"os"
+	"path"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/zrepl/yaml-config"
+)
+
+func TestTLSConfig(t *testing.T) {
+	input := `
+cert: path/to/cert.pem
+key: path/to/key.pem
+client_ca: path/to/ca.pem
+client_auth_type: RequireAndVerifyClientCert
+min_version: TLS1.0
+max_version: TLS1.3
+cipher_suites:
+  # TLS 1.2 cipher suites.
+  - TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA
+  - TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA
+  - TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA
+  - TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA
+  - TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
+  - TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
+  - TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
+  - TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
+  - TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
+  - TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256
+  # TLS 1.3 cipher suites.
+  - TLS_AES_128_GCM_SHA256
+  - TLS_AES_256_GCM_SHA384
+  - TLS_CHACHA20_POLY1305_SHA256
+`
+	parsed := TLSConfig{
+		CertPath:   "path/to/cert.pem",
+		KeyPath:    "path/to/key.pem",
+		ClientCAs:  "path/to/ca.pem",
+		ClientAuth: ClientAuthType(tls.RequireAndVerifyClientCert),
+		MinVersion: tls.VersionTLS10,
+		MaxVersion: tls.VersionTLS13,
+		CipherSuites: []CipherSuite{
+			// TLS 1.2 cipher suites.
+			CipherSuite(tls.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA),
+			CipherSuite(tls.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA),
+			CipherSuite(tls.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA),
+			CipherSuite(tls.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA),
+			CipherSuite(tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256),
+			CipherSuite(tls.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384),
+			CipherSuite(tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256),
+			CipherSuite(tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384),
+			CipherSuite(tls.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256),
+			CipherSuite(tls.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256),
+			// TLS 1.3 cipher suites.
+			CipherSuite(tls.TLS_AES_128_GCM_SHA256),
+			CipherSuite(tls.TLS_AES_256_GCM_SHA384),
+			CipherSuite(tls.TLS_CHACHA20_POLY1305_SHA256),
+		},
+	}
+
+	var actual TLSConfig
+	err := yaml.Unmarshal([]byte(input), &actual)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !reflect.DeepEqual(actual, parsed) {
+		t.Fatal("parsed TLS config doesn't match expected")
+	}
+
+	workdir := t.TempDir()
```

Comment:
This is a Golang 1.15 feature, but I notice that this project is still stuck on the ancient Go 1.12 (not sure why, but I tried my best within the constraints).
I assumed that this was fine to break that constraint given that this is just test code